pii/ner_model_training/utils/chunking.py

```diff
@@ -0,0 +1,187 @@
+import itertools
+from typing import List, Iterable, Any, Optional
+
+import numpy as np
+from datasets import Dataset
+from scipy.special import softmax
+from tqdm.notebook import tqdm
+from transformers import PreTrainedTokenizer
+from transformers.trainer_utils import PredictionOutput
+
+from utils.misc import LABEL2ID
+
+
+def chunk_dataset(dataset: Dataset, tokenizer: PreTrainedTokenizer, overlap_freq: int = 0) -> Dataset:
+    """
+    Splits every entry in the `dataset` into chunks of the specific length.
+    The maximum length of each chunk is set by the max_len_single_sentence attribute of the `tokenizer`.
+    The `overlap_freq` argument controls the amount of overlap between adjacent chunks.
+    If overlap_freq is 0, there is no overlap between chunks.
+    If overlap_freq is 1, adjacent chunks overlap by half of their length.
+    If overlap_freq is greater than 2, the overlap is set to a `1. / overlap_freq` fraction.
+
+    The resulting dataset contains the following columns: `input_ids`, `attention_mask`, `labels`, `id`, and `chunk_id`.
+    Each row in the dataset corresponds to a single chunk of the original data, identified by its `id` and `chunk_id`.
+    """
+    return Dataset.from_list(
+        list(
+            itertools.chain(
+                *(chunk_inputs(**entry, tokenizer=tokenizer, max_length=tokenizer.max_len_single_sentence,
+                               overlap_freq=overlap_freq) for entry in tqdm(list(dataset)))
+            )
+        )
+    )
+
+
+def chunk_inputs(input_ids,
+                 attention_mask,
+                 id,
+                 tokenizer,
+                 max_length,
+                 labels=None,
+                 overlap_freq=0,
+                 **kwargs):
+    """
+    Processes a sequence of `input_ids`, `attention_mask`, and `labels` by dividing it
+    into smaller chunks to be processed by the model. The sequence is split into chunks
+    of a specified maximum length, with a defined overlap frequency `overlap_freq`. The function
+    handles special tokens that should be added to the beginning and end of the chunks.
+
+    Returns
+    -------
+    list of dictionaries that contain the chunked `input_ids`, `attention_mask`, `labels`
+
+    """
+
+    step = _get_chunking_step(max_length, overlap_freq)
+
+    def _chunked_seq(seq):
+        for i in range(len(seq) // step + 1):
+            if i * step < len(seq):
+                yield seq[i * step:i * step + max_length]
+
+    if labels is not None:
+        chunks = zip(*[_chunked_seq(seq) for seq in (input_ids, attention_mask, labels)])
+        chunks = (_prepare_for_model(*chunk, tokenizer=tokenizer) for chunk in chunks)
```

Comment:
we should probably add the special tokens to beginning and end of each file before chunking